workspace/src/localization/localization_py/localization_py/particle_filter.py

```diff
@@ -0,0 +1,180 @@
+import numpy as np
+import math
+import matplotlib.pyplot as plt
+import random
+from scipy.stats import wasserstein_distance
+import yaml
+from dynamics import Dynamics
+
+class ParticleFilter:
+    """
+    A basic Particle Filter implementation.
+    
+    This currently uses the 4DOF motion model. The Motion Model inputs are throttle and steering, and the observation model uses GPS and Magnetometer as sensors. States are cartesian coordinates x, y, heading theta, and velocity v. Further reference can be found at: https://sbel.wisc.edu/wp-content/uploads/sites/569/2023/07/TR-2023-05.pdf.
+
+    Attributes:
+        show_animation: Boolean for whether or not to show the particle animation.
+        dt: The timestep for the filter.
+        num_particles: The number of particles maintained by the filter.
+        particles: The array of particles.
+        hpx: The array of x coordinate histories for each particle. Used for the animation.
+        hpy: The array of y coordinate histories for each particle. Used for the animation.
+        particle_distr_dist: The expected noise distribution for the position of the particles.
+        particle_distr_head: The expected noise distribution for the heading of the particles.
+        timestep: the current timestep. Used for the animation.
+    
+    """
+
+    def __init__(self, dt, dynamics):
+        """Initialize the PF.
+
+        Initialize each of the global variables for the PF. Load the dynamics parameters for the vehicle from an external YAML file.
+
+        Args:
+            dt: The timestep at which this filter will operate at. 
+            dynamics: the parameters for the dynamics
+        """
+        self.dt = dt
+        self.dyn = Dynamics(dt, dynamics)
```

Comment:
Same dt problem pointed out in EKF class

Reply:
@StefanCaldararu  Would it be reasonable to make this dt related with the frequency of the state estimation node?

Reply:
@harryzhang1018 i'm confused, that's the way it is written here right? and that's what Huzafia doesn't like